readme
1) 权限控制方案为：
If( 页面添加了<accesscontrol>权限控制){
     If( <accesscontrol>标签中有_anonymous_ ){
              允许访问该页面(不登陆可只读访问);
     }
     If(<accesscontrol>标签中有当前登陆用户所在组){
              允许当前用户访问该页面;
     }else{
              禁止当前用户访问该页面;
     }
}else{
     允许访问该页面(不登陆可只读访问);
}

2) 标签格式：<accesscontrol>bizop,bizqa</accesscontrol>


changelog.txt
version 3.4
* modified for sogou.biztech

version 2.1
* Fixed security problem with insert protected page, or template

version 2.0
* Completly rewrited for MediaWiki >= 1.18 by Aleš Kapica
* Removed not used variables:
    $wgAccessControlDebug
    $wgAccessControlDebugFile
    $wgAllowInfo
    $wgAllowUserList
    $wgAccessControlMessages

version 1.3
* Gabor Simon fixed problem in function getUsersFromPages()
* Applying patch from Stephan Herrmann

version 1.2
* Fixed errors in parameter fuction doControlUserAccess()
* Repairing while lopp in function getContentTag()

version 1.1
* Fixed security bug with accessibility on protected page over #REDIRECT

version 1.0
* Refactored all code
* Added localization file
* Changed separator on one comma (from double)
* Readonly access is can not only group, but user too

version 0.8
*Fixes for hook errors in MediaWiki 1.11.0 included
*Support for the Group of anonymous users added
*UNIQ/QUINU-Bug kind of fixed (own Link-Parser)

version 0.7
*now it's possible to have read-only groups.
*set the english language as default
*refactored some functions for better readability
*it's now possible to use the internal groups from MediaWiki
*Added a TODO-Document

version 0.6
*Bugfix release, touching the article didn't work, so I use a header redirect (bad hack, but works).
*works now also for anonymous users

version 0.5
*Articles with the accesscontrol tag are now touched in advance to displaying it, so the page will not read from the cache
*removed some development junk ;-)
*better documentation of the accesscontrolSettings.php
*added some debugging code

version 0.4
*tested in Version 1.8.2
*added some Tips
*if more then one group is in the <accesscontrol> tag, all groups are now displayed
*there is now an option, where you can turn off the behaviour, that users in the sysop group can see restricted pages
*the "access is allowed only for group..." text is now in the accesscontrolSettings.php for easier localisation

version 0.3
*access is now also controlled for editing the pages (if you access it manually per action=edit in the URL)
*only sysops can now view and edit the "Usergroup:.." pages
*added a changelog ;-)

version 0.2
*some fixes, to make it work under Version 1.7.1
*Sysops can now always access pages, so if you make an error, you have always the opportunity to correct it ;-)

version 0.1
first Version


AccessControl.php
<?php

/* MediaWiki extension that enables group access restriction on a page-by-page
 * basis contributed by Martin Mueller (http://blog.pagansoft.de) based into 
 * version 1.3 on accesscontrol.php by Josh Greenberg.
 * Version 2.0 for MediaWiki >= 1.18 rewrited completly by Aleš Kapica.
 * Version 3.4 for MediaWiki >= 1.19 . Modified by shanhongjie@sogou-inc.com in order to used it independent on the group page.
 * Usage:<accesscontrol>biztech, bizop </accesscontrol>
 * @package MediaWiki
 * @subpackage Extensions
 * @author Aleš Kapica
 * @copyright 2008-2012 Aleš Kapica
 * @modified by shanhongjie@sogou-inc.com,2012-12-24
 * @licence GNU General Public Licence
 */

if( !defined( 'MEDIAWIKI' ) ) {
	echo ( "This file is an extension to the MediaWiki software and cannot be used standalone.\n" );
	die();
}

// sysop users can read all restricted pages
$wgAdminCanReadAll = true;

$wgExtensionCredits['specialpage']['AccessControl'] = array(
	'name'                  => 'AccessControlExtension',
	'author'                => array( 'shanhongjie@sogou-inc.com' ),
	'url'                   => 'git://github.com/ujfjhz/mediawiki-extensions-AccessControl.git',
	'version'               => '3.4',
	'description'           => 'Access control based on users with group info. Administrator rights need not be for it.',
	'descriptionmsg'        => 'accesscontrol-desc',
);

$wgHooks['ParserFirstCallInit'][] = 'wfAccessControlExtension' ;

$dir = dirname( __FILE__ ) . '/';
$wgExtensionMessagesFiles['AccessControl'] = $dir . 'AccessControl.i18n.php';


//Hook the userCan function for bypassing the cache
$wgHooks['userCan'][] = 'hookUserCan';


function wfAccessControlExtension( Parser $parser ) {
	/* This the hook function adds the tag <accesscontrol> to the wiki parser */
	$parser->setHook( "accesscontrol", "doControlUserAccess" );
	return true;
}

function doControlUserAccess( $input, array $args, Parser $parser, PPFrame $frame ) {
	/* Funcion called by wfAccessControlExtension */
	return displayGroups();
}

#obsahtagu(the content of tag(slovak))
function accessControl( $obsahtagu ){
	$accessgroup = Array( Array(), Array() );
	$listaccesslist = explode( ",", $obsahtagu );
	foreach ( $listaccesslist as $accesslist ) {
		if ( strpos( $accesslist, "(ro)" ) !== false ) {
			$accesslist = trim( str_replace( "(ro)", "", $accesslist ) );
			$group = makeGroupArray( $accesslist );
			$accessgroup[1] = array_merge( $accessgroup[1], $group[0] );
			$accessgroup[1] = array_merge( $accessgroup[1], $group[1] );
		} else {
			$accesslist = trim( $accesslist );
			$group = makeGroupArray ($accesslist );
			$accessgroup[0] = array_merge( $accessgroup[0], $group[0] );
			$accessgroup[1] = array_merge( $accessgroup[1], $group[1] );
		}
	}
	return $accessgroup;
}

function makeGroupArray( $accesslist ) {
	/* Function returns array with two lists.
		First is list full access users.
		Second is list readonly users. */
	$userswrite = Array();
	$usersreadonly = Array();
	$users = getUsersFromPages( $accesslist );
	foreach ( array_keys( $users ) as $user ) {
		switch ( $users[$user] ) {
			case 'read':
				$usersreadonly[] = $user;
				break;
			case 'edit':
				$userswrite[] = $user;
				break;
		}
	}
	return array( $userswrite , $usersreadonly );
}

function displayGroups() {
	/* Function replace the tag <accesscontrol> and his content, behind info about a protection this the page */
	$style = "<p id=\"accesscontrol\" style=\"text-align:center;color:#BA0000;font-size:8pt\">";
	$text = wfMsg( 'accesscontrol-info' );
	$style_end = "</p>";
	$wgAllowInfo = $style . $text . $style_end;
	return $wgAllowInfo;
}

function getContentPage( $title ) {
	/* Function get content the page identified by title object from database */
	$Title = new Title();
	$gt = $Title->makeTitle( 0, $title );
	// create Article and get the content
	$contentPage = new Article( $gt, 0 );
	return $contentPage->fetchContent( 0 );
	}

function getTemplatePage( $template ) {
	/* Function get content the template page identified by title object from database */
	$Title = new Title();
	$gt = $Title->makeTitle( 10, $template );
	//echo '<!--';
	//print_r($gt);
	//echo '-->';
	// create Article and get the content
	$contentPage = new Article( $gt, 0 );
	return $contentPage->fetchContent( 0 );
	}

function getUsersFromPages( $skupina ) {
	/* Extracts the allowed users from the userspace access list */
	$allowedAccess = Array();
	$allow = Array();
	$Title = new Title();
	$gt = $Title->makeTitle( 0, $skupina );
	// create Article and get the content
	$groupPage = new Article( $gt, 0 );
	$allowedUsers = $groupPage->fetchContent( 0 );
	$groupPage = NULL;
	$usersAccess = explode( "\n", $allowedUsers );
	foreach  ($usersAccess as $userEntry ) {
		$userItem = trim( $userEntry );
		if ( substr( $userItem, 0, 1 ) == "*" ) {
			if ( strpos( $userItem, "(ro)" ) === false ) {
				$user = trim( str_replace( "*", "", $userItem ) );
				$allow[$user] = 'edit';
			} else {
				$user = trim( str_replace( "*", "", $userItem ) );
				$user = trim( str_replace( "(ro)", "", $user ) );
				$allow[$user] = 'read';
			}
		}
	}
	if ( is_array( $allow ) ) {
		$allowedAccess = $allow;
		unset( $allow );
	}
	return $allowedAccess;
}

function doRedirect( $info ) {
	/* make redirection for non authorized users */
	global $wgScript, $wgSitename, $wgOut;

	if ( ! $info ) {
	    $info = "No_access";
	    }
	if ( $info == "Only_sysop" ) {
		$target = wfMsg( 'accesscontrol-info-user' );
	} elseif ( $info == "No_anonymous" ) {
		$target = wfMsg( 'accesscontrol-info-anonymous' );
	} elseif ( $info == "Deny_anonymous") {
		$target = wfMsg( 'accesscontrol-edit-anonymous' );
	} elseif ( $info == "Deny_edit_list" ) {
		$target = wfMsg( 'accesscontrol-edit-users' );
	} else {
		$target = wfMsg( 'accesscontrol-info-deny' );
	}
	if ( isset( $_SESSION['redirect'] ) ) {
		// removing info about redirect from session after move..
		unset( $_SESSION['redirect'] );
	}
	header( "Location: " . $wgScript . "/" . $wgSitename . ":" . $target );
}

function fromTemplates( $string ) {
	global $wgUser, $wgAdminCanReadAll;
	// Vytažení šablon
	if ( strpos( $string, '{{' ) ) {
	    if ( substr( $string, strpos ( $string, '{{' ), 3 ) === '{{{' ) {
		    $start = strpos( $string, '{{{' );
		    $end = strlen( $string );
		    $skok = $start + 3;
		    fromTemplates( substr( $string, $skok, $end - $skok ) );
		} else {
		    $start = strpos( $string, '{{' );
		    $end = strpos( $string, '}}' );
		    $skok = $start + 2;
		    $templatepage = substr( $string, $skok, $end - $skok );
		    if ( strpos( $templatepage, '|' ) > 0) { 
			    $templatename = substr( $templatepage, 0, strpos( $templatepage, '|' ) );
			} else {
			    $templatename = $templatepage ;
			}
		    if ( substr( $templatename, 0, 1 ) === ':') {
			    // vložena stránka
			    $rights = allRightTags( getContentPage( substr( $templatename, 1 ) ) );
			} else {
			    // vložena šablona
			    $rights = allRightTags( getTemplatePage( $templatename ) );
			}
		    if ( is_array( $rights ) ) {
			if ( $wgUser->mId === 0 ) {
			    /* Redirection unknown users */
			    $wgActions['view'] = false;
			    doRedirect('accesscontrol-info-anonymous');
			    } else {
				if ( in_array( 'sysop', $wgUser->mGroups, true ) ) {
					if ( isset( $wgAdminCanReadAll ) ) {
						if ( $wgAdminCanReadAll ) {
							return true;
							}
						}
					}
				$users = accessControl( $rights['groups'] );
				if ( ! in_array( $wgUser->mName, $users[0], true ) ) {
					$wgActions['edit']           = false;
					$wgActions['history']        = false;
					$wgActions['submit']         = false;
					$wgActions['info']           = false;
					$wgActions['raw']            = false;
					$wgActions['delete']         = false;
					$wgActions['revert']         = false;
					$wgActions['revisiondelete'] = false;
					$wgActions['rollback']       = false;
					$wgActions['markpatrolled']  = false;
					if ( ! in_array( $wgUser->mName, $users[1], true ) ) {
						$wgActions['view']   = false;
						return doRedirect( 'accesscontrol-info-anonymous' );
						}
					}
				}
			}
		    fromTemplates( substr( $string, $end + 2 ) );
		}
	    }
    }


function allRightTags( $string ) {
	/* Function for extraction content tag accesscontrol from raw source the page */
	$contenttag  = Array();
	$starttag    = "<accesscontrol>";
	$endtag      = "</accesscontrol>";
	$redirecttag = "redirect";

	if ( ( mb_substr( trim( $string ), 0, 1 ) == "#" )
		&& ( stripos( mb_substr( trim( $string ), 1, 9 ), $redirecttag ) == "0" )
		) {
		/* Treatment redirects - content variable $string must be replaced over content the target page */
		$sourceredirecttag = mb_substr( $string, 0, strpos( $string, ']]' ) );
		$redirecttarget = trim( substr( $sourceredirecttag, strpos( $sourceredirecttag, '[[' ) + 2 ) );
		if ( strpos( $redirecttarget, '|' ) ) {
			$redirecttarget = trim( substr( $redirecttarget, 0, strpos( $redirecttarget, '|' ) ) );
		}
		$Title = new Title();
		$gt = $Title->makeTitle( 0, $redirecttarget );
		return allRightTags( getContentPage( $gt ) );
	}
	// Kontrola accesscontrol ve vložených šablonách a stránkách
	fromTemplates($string);

	$start = strpos( $string, $starttag );
	if ( $start !== false ) {
		$start += strlen( $starttag );
		$end = strpos( $string, $endtag );
		if ( $end !== false ) {
			$groupsString = substr( $string, $start, $end-$start );
			if ( strlen( $groupsString ) == 0 ) {
				$contenttag['end'] = strlen( $starttag ) + strlen( $endtag ); 
			} else {
				$contenttag['groups'] = $groupsString;
				$contenttag['end'] = $end + strlen( $endtag );
			}

			if( isset( $_SESSION['redirect'] ) ) {
				$_SESSION['redirect'] = $contenttag;
			} else {
				return $contenttag;
			}
		}
	} else {
		if( isset( $_SESSION['redirect'] ) ) {
			return $_SESSION['redirect'];
		} else {
			return false;
		}
	}
}

function hookUserCan( &$title, &$wgUser, $action, &$result ) {
	/* Main function control access for all users */
	global $wgActions, $wgAdminCanReadAll;
	$rights = allRightTags( getContentPage( $title->mDbkeyform ) );

	if ( is_array( $rights ) ) {
#added by shanhongjie@sogou-inc.com
	$restrictGrouplist=explode(',',$rights['groups']);
	foreach($restrictGrouplist as $restrictGroup){
		if ( $restrictGroup == "_anonymous_") {
			return true;
		}
	}

		if ( $wgUser->mId === 0 ) {
			/* Redirection unknown users */
			$wgActions['view'] = false;
			doRedirect( 'accesscontrol-info-anonymous' );
		} else {
			if ( in_array( 'sysop', $wgUser->mGroups, true ) ) {
				if ( isset( $wgAdminCanReadAll ) ) {
					if ( $wgAdminCanReadAll ) {
						return true;
					}
				}
			}				

	foreach($restrictGrouplist as $restrictGroup){
		//TODO ro,rw,etc
		if ( in_array( trim($restrictGroup),$wgUser->mGroups, true ) ) {
			return true;
		}
	}
	$wgActions['view'] = false;
	doRedirect( 'accesscontrol-info-anonymous' );
		}
	} else {
#		if ( $wgUser->mId === 0 && $title->mNamespace === 0 ) {
#			/* Redirection:when unknown users in 0 mNamespace without _anonymous_ promise */
#			$wgActions['view'] = false;
#			doRedirect( 'accesscontrol-info-anonymous' );
#		}else{
#			return true;
#		}
		return true;
	}
}

?>
